opendir.c

```diff
@@ -47,23 +47,91 @@ struct watchman_dir_handle {
   struct watchman_dir_ent ent;
 };
 
+/* Opens a directory, strictly prohibiting opening any symlinks
+ * in any component of the path on systems that make this
+ * reasonable and easy to do.
+ * We do this by opening / and then walking down the tree one
```

Comment:
We could, but I think this way is slightly less prone to TOCTOU and is possibly more efficient in case any element of the path is a symlink; we'll bail immediately, whereas the realpath approach would resolve all the way down before we got to compare the names.